### PR TITLE
Create program for ADSL dataset

### PR DESCRIPTION
**Pull request by SHAESEN2**
#1

### CODE REVIEW

File: submission/programs/adsl.R [hunk 1]
```diff
@@ -0,0 +1,455 @@
...
+  dplyr::select(STUDYID, USUBJID, VISIT1DT)
+
+visnumen <- sv %>%
+  dplyr::filter(VISITNUM < 100) %>%
+  dplyr::arrange(STUDYID, USUBJID, SVSTDTC) %>%
+  dplyr::group_by(STUDYID, USUBJID) %>%
+  dplyr::slice(n()) %>%
+  dplyr::ungroup() %>%
+  dplyr::mutate(VISNUMEN = ifelse(round(VISITNUM, digits = 0) == 13, 12, round(VISITNUM, digits = 0))) %>%
+  dplyr::select(STUDYID, USUBJID, VISNUMEN)
```
Comment (DeclanHodges) [inline submission/programs/adsl.R:321]: According to the SAP, this derivation should be done using DS:
image

**Review by DeclanHodges — CHANGES_REQUESTED** [review]
Hi Steven,
Thanks for putting this code together, it's looking good so far! I've attached some comments to my review, please let me know what you think.

File: submission/programs/adsl.R [hunk 2]
```diff
@@ -0,0 +1,455 @@
...
+mh <- admiral::convert_blanks_to_na(haven::read_xpt(file.path("sdtm", "mh.xpt")))
+
+adsl_prod <- admiral::convert_blanks_to_na(haven::read_xpt(file.path(adam[2], "adsl.xpt")))
+
+toprogram <- setdiff(colnames(adsl_prod), colnames(dm))
+
+# Formats -----------------------------------------------------------------
+
+# site groups - if not pooled then SITEGR1=SITEID. If pooled, SITEGR1 will be 900 - no SAP available
+format_siteid <- function(x) {
```
[the remark this answers is not in the document]
Comment (thomas-neitmann) [reply submission/programs/adsl.R:40]: I think that would be good.

**Comment by DeclanHodges** [comment]
Also, should this branch be being merged into adsl rather than main?

**Review by thomas-neitmann — CHANGES_REQUESTED** [review]
Two general comments:

1. Please don't use `pkg::fun()` but just `fun()`
2. Please move all custom functions you define in the `R` folder and add `library(pilot3)` at the top of your script to make these functions available. I'd put all `format_*()` functions into `R/formats.R`.

**Comment by SHAESEN2** [comment]
> Two general comments:
> 
> 
> 
> 1. Please don't use `pkg::fun()` but just `fun()`
> 
> 2. Please move all custom functions you define in the `R` folder and add `library(pilot3)` at the top of your script to make these functions available. I'd put all `format_*()` functions into `R/formats.R`.

1. I disagree since I had issues before eg filter and lag. I think it helps with understanding and debugging for programmers not that familiar with R. Me as well, I often need to figure out from which package a function is to understand issues in the program.

2. I disagree. The formats can live in the same program as they are only used once. Adding a folder formats makes it more difficult to submit and explain in ADRG and makes it more difficult to debug and understand for more junior R programmers. For library(pilot 3) i agree completely.

File: submission/programs/adsl.R [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+    x == "BLACK OR AFRICAN AMERICAN" ~ 2,
+    x == "AMERICAN INDIAN OR ALASKA NATIVE" ~ 6
+  )
+}
+# BMI group
+format_bmi <- function(x) {
+  dplyr::case_when(
+    !is.na(x) & x < 25 ~ "<25",
+    25 <= x & x < 30 ~ "25-<30",
+    30 <= x ~ ">=30"
```
[the remark this answers is not in the document]
Comment (SHAESEN2) [reply submission/programs/adsl.R:77]: I disagree. Define needs an update ro match actual dataset values.

File: submission/programs/adsl.R [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+adsl_prod <- admiral::convert_blanks_to_na(haven::read_xpt(file.path(adam[2], "adsl.xpt")))
+
+toprogram <- setdiff(colnames(adsl_prod), colnames(dm))
+
+# Formats -----------------------------------------------------------------
+
+# site groups - if not pooled then SITEGR1=SITEID. If pooled, SITEGR1 will be 900 - no SAP available
+format_siteid <- function(x) {
+  dplyr::case_when(
+    x %in% c("702", "706", "707", "711", "714", "715", "717") ~ "900",
```
[the remark this answers is not in the document]
Comment (SHAESEN2) [reply submission/programs/adsl.R:42]: I followed the logic in the actual dataset. Thanks for pointing this out. Ill double check with SAP. Can you make it available or did you mean protocol?

File: submission/programs/adsl.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (SHAESEN2) [reply submission/programs/adsl.R:40]: I agree. Should be updated.

File: submission/programs/adsl.R [hunk 5]
```diff
@@ -0,0 +1,455 @@
...
+    new_vars = vars(EOSDT = EOSDT),
+    filter_add = DCDECOD != "COMPLETED"
+  ) %>%
+  admiral::derive_vars_dt(
+    dtc = EXENDTC,
+    new_vars_prefix = "EXEN",
+    highest_imputation = "Y",
+    min_dates = vars(EXSTDT), 
+    max_dates = vars(EOSDT),
+    date_imputation = "last"
```
[the remark this answers is not in the document]
Comment (SHAESEN2) [reply submission/programs/adsl.R:128]: Good point!

File: submission/programs/adsl.R [hunk 6]
```diff
@@ -0,0 +1,455 @@
...
+    RACEN = format_racen(RACE)
+  )
+
+# Population flag ---------------------------------------------------------
+# SAFFL - Y if ITTFL='Y' and TRTSDT ne missing. N otherwise
+# ITTFL - Y if ARMCD ne ' '. N otherwise
+# EFFFL - Y if SAFFL='Y AND at least one record in QS for ADAS-Cog and for CIBIC+ with VISITNUM>3, N otherwise
+# these variables are also in suppdm, but define said derived
+
+qstest <- distinct(qs[,c("QSTESTCD", "QSTEST")])
```
[the remark this answers is not in the document]
Comment (SHAESEN2) [reply submission/programs/adsl.R:203]: Should not be an issue for submission or readability

File: submission/programs/adsl.R (hunk 1 above)
Comment (SHAESEN2) [reply submission/programs/adsl.R:321]: This is the define you show. I saw but this information is not in DS. I used DS.DSSTDTC information which should be sufficient.

**Comment by SHAESEN2** [comment]
> Hi Steven,
> 
> Thanks for putting this code together, it's looking good so far! I've attached some comments to my review, please let me know what you think.

Good comment! Thanks for the review. I will make time next week for the updates unless someone else wants to try.

We should merge with main once finalized unless adsl branch has code inside. Then we can merge.

File: submission/programs/adsl.R (hunk 4 above)
[the remark this answers is not in the document]
Comment (DeclanHodges) [reply submission/programs/adsl.R:42]: The SAP is contained within the CSR. You can find it here in the Slack channel: https://pharmaverse.slack.com/archives/C04968ES7T4/p1668184405997029. The screen shot I've taken is of page 425.

File: submission/programs/adsl.R [hunk 7]
```diff
@@ -0,0 +1,455 @@
...
+
+
+dm <- admiral::convert_blanks_to_na(haven::read_xpt(file.path("sdtm", "dm.xpt")))
+ds <- admiral::convert_blanks_to_na(haven::read_xpt(file.path("sdtm", "ds.xpt")))
+ex <- admiral::convert_blanks_to_na(haven::read_xpt(file.path("sdtm", "ex.xpt")))
+qs <- admiral::convert_blanks_to_na(haven::read_xpt(file.path("sdtm", "qs.xpt")))
+sv <- admiral::convert_blanks_to_na(haven::read_xpt(file.path("sdtm", "sv.xpt")))
+vs <- admiral::convert_blanks_to_na(haven::read_xpt(file.path("sdtm", "vs.xpt")))
+sc <- admiral::convert_blanks_to_na(haven::read_xpt(file.path("sdtm", "sc.xpt")))
+mh <- admiral::convert_blanks_to_na(haven::read_xpt(file.path("sdtm", "mh.xpt")))
```
[the remark this answers is not in the document]
Comment (SHAESEN2) [reply submission/programs/adsl.R:31]: I would leave as is. We could have a small function that reads one sdtm dataset at a time, but storing the datasets in a list is overkill.

File: submission/programs/adsl.R (hunk 4 above)
[the remark this answers is not in the document]
Comment (SHAESEN2) [reply submission/programs/adsl.R:42]: dm %>%
  dplyr::group_by(SITEID, ACTARMCD) %>%
  dplyr::summarise(n=n_distinct(USUBJID)) %>%
  dplyr::filter(n<3, ACTARMCD != "Scrnfail") %>%
  dplyr::summarise(n_distinct(SITEID))
  
```r
  SITEID `n_distinct(SITEID)`
  <chr>                 <int>
1 702                       1
2 706                       1
3 707                       1
4 711                       1
5 714                       1
6 715                       1
7 717                       1
```

File: submission/programs/adsl.R [hunk 8]
```diff
@@ -0,0 +1,455 @@
...
+
+
+#dfcompare(
+#  file = "adsl_compare"
+#  ,left = adsl_prod
+#  ,right = adsl
+#  ,keys = c("STUDYID", "USUBJID")
+#  ,showdiffs = 10000
+#  ,debug = F
+#)
```
[the remark this answers is not in the document]
Comment (SHAESEN2) [reply submission/programs/adsl.R:456]: done

**Review by DeclanHodges — APPROVED** [review]

**Comment by thomas-neitmann** [comment]
To fix the failing workflows please add `^submission$` to `.Rbuildignore` file and run `styler::style_file("submission/programs/adsl.R")`.

**Comment by thomas-neitmann** [comment]
> 1. I disagree since I had issues before eg filter and lag. I think it helps with understanding and debugging for programmers not that familiar with R. Me as well, I often need to figure out from which package a function is to understand issues in the program.

I don't think that is an issue as R clearly tells you when a function from a package masks another function with the same name from another package. If you want to be cautious you could use the `pkg::` prefix for just the functions which have conflicts, e.g. `dplyr::filter()` but using the `pkg::` for all functions just makes the program more hard to read. By typing `?fun` the help page will tell you from which package the function comes in case you don't know.

**Comment by SHAESEN2** [comment]
> To fix the failing workflows please add `^submission$` to `.Rbuildignore` file and run `styler::style_file("submission/programs/adsl.R")`.

Done.

**Comment by SHAESEN2** [comment]
> > 2. I disagree. The formats can live in the same program as they are only used once. Adding a folder formats makes it more difficult to submit and explain in ADRG and makes it more difficult to debug and understand for more junior R programmers. For library(pilot 3) i agree completely.
> 
> We want to simulate a realistic submission. Typically you'd have more than a single study for a particular molecule and would re-use function across studies. Thus, it makes sense to put the functions in an R package which is exactly the approach the pilot 1 team has followed. From what I've seen many of the functions they created are also used just once. So to re-iterate, please mive all the `format_*()` functions inside the `R` folder such that they become part of the `pilot3` package which you can then load at the top of your script.

I see your point assuming that all studies are perfectly aligned with variable names. I think creating a package for submitting functions is interesting but is it workable in reality? This approach implies we need to train all lead programmers to create a package, our computing environment must support package creation using subfolders and accept the different extension to move from development to a production environment, .... 
I would propose to use pkglite to create the txt without relying on the package structure. It is perfectly feasible, much more simple and easier to explain to reviewers.

**Comment by SHAESEN2** [comment]
> > 1. I disagree since I had issues before eg filter and lag. I think it helps with understanding and debugging for programmers not that familiar with R. Me as well, I often need to figure out from which package a function is to understand issues in the program.
> 
> I don't think that is an issue as R clearly tells you when a function from a package masks another function with the same name from another package. If you want to be cautious you could use the `pkg::` prefix for just the functions which have conflicts, e.g. `dplyr::filter()` but using the `pkg::` for all functions just makes the program more hard to read. By typing `?fun` the help page will tell you from which package the function comes in case you don't know.

I understand the concern. However, based off my experience with other new R programmers, I prefer to keep it as it helps them to understand the programs easier. Selecting ?fun every time for a large code is very time consuming for new programmers.

**Comment by SHAESEN2** [comment]
Hey @DeclanHodges I think we should first finish the above discussion before making changes that might not be required. I don't think making a pilot 3 package for this effort is required so the R functions can stay where they are.

**Comment by DeclanHodges** [comment]
> Hey @DeclanHodges I think we should first finish the above discussion before making changes that might not be required. I don't think making a pilot 3 package for this effort is required so the R functions can stay where they are.

Hi @SHAESEN2, Thomas has asked me to go ahead and make these updates before he's happy to merge this PR.

**Review by thomas-neitmann — APPROVED** [review]
Thanks a lotf or your efforts @SHAESEN2 and your support @DeclanHodges 👏

**Comment by DeclanHodges** [comment]
> Thanks a lotf or your efforts @SHAESEN2 and your support @DeclanHodges 👏

No problem, thanks for your help too @thomas-neitman!

**Pull request merged**